vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/jsonitem/SimpleObjectUnionScalarEvaluator.java

```diff
@@ -44,60 +44,58 @@ public SimpleObjectUnionScalarEvaluator(IHyracksTaskContext ctx, IScalarEvaluato
+        TaggedValuePointable tempTvp = ppool.takeOne(TaggedValuePointable.class);
+        TaggedValuePointable tempValue = ppool.takeOne(TaggedValuePointable.class);
+        UTF8StringPointable tempKey = ppool.takeOne(UTF8StringPointable.class);
+        try {
+            abvs.reset();
+            ob.reset(abvs);
+            tvps.clear();
+            TaggedValuePointable arg = args[0];
+            if (arg.getTag() == ValueTag.SEQUENCE_TAG) {
+                arg.getValue(sp);
                 for (int i = 0; i < sp.getEntryCount(); ++i) {
                     op = (ObjectPointable) ObjectPointable.FACTORY.createPointable();
                     sp.getEntry(i, tempTvp);
                     tempTvp.getValue(op);
                     op.getKeys(tempTvp);
                     if (tempTvp.getTag() == ValueTag.XS_STRING_TAG) {
-                        key = ppool.takeOne(TaggedValuePointable.class);
-                        value = ppool.takeOne(TaggedValuePointable.class);
-                        tempTvp.getValue(tempKey);
-                        op.getValue(tempKey, value);
-                        key.set(tempTvp);
-                        tvps.add(key);
-                        tvps.add(value);
-                        tvps.add(boolTvp);
-
+                        addPair(tempTvp, tempKey, tempValue);
                     } else if (tempTvp.getTag() == ValueTag.SEQUENCE_TAG) {
                         tempTvp.getValue(sp1);
                         for (int j = 0; j < sp1.getEntryCount(); ++j) {
                             key = ppool.takeOne(TaggedValuePointable.class);
-                            value = ppool.takeOne(TaggedValuePointable.class);
                             sp1.getEntry(j, tempTvp);
-                            tempTvp.getValue(tempKey);
-                            op.getValue(tempKey, value);
-                            key.set(tempTvp);
-                            tvps.add(key);
-                            tvps.add(value);
-                            tvps.add(boolTvp);
+                            addPair(tempTvp, tempKey, tempValue);
                         }
-
                     }
                 }
-                super.evaluate(tvps.toArray(new TaggedValuePointable[tvps.size()]), result);
-            } catch (IOException e) {
-                throw new SystemException(ErrorCode.SYSE0001, e);
-            } finally {
-                ppool.giveBack(tempKey);
-                ppool.giveBack(tempTvp);
-                ppool.giveBack(boolTvp);
-                for (TaggedValuePointable pointable : tvps) {
-                    ppool.giveBack(pointable);
-                }
+            }
+            ob.finish();
+            result.set(abvs);
+        } catch (IOException e) {
+            throw new SystemException(ErrorCode.SYSE0001, e);
+        } finally {
+            ppool.giveBack(tempKey);
+            ppool.giveBack(tempTvp);
+            for (TaggedValuePointable pointable : tvps) {
```

Comment:
Create a shared method (in the abstract class) for returning all tvps pointables to the ppool.